packages/services/src/providers/price/coingecko.provider.ts

```diff
@@ -0,0 +1,432 @@
+   * - The timestamp is natively in ISO 8601 format and must be converted to unix.
+   * @param pools - The token pools for the token
+   * @returns The created at timestamp
+   */
+  private getCreatedAtTimestampFromPools(
+    pools: TokenGetAddressResponse.Included[],
+  ): number | undefined {
+    const poolsWithTimestamps = pools.filter(
+      (p) => p.attributes?.pool_created_at,
+    );
+    if (poolsWithTimestamps.length === 0) {
+      return undefined;
+    }
+    // Note: the timestamps are guaranteed to be non-null because we filtered above
+    const sorted = poolsWithTimestamps.sort((a, b) => {
+      const aTime = new Date(a.attributes!.pool_created_at!).getTime();
+      const bTime = new Date(b.attributes!.pool_created_at!).getTime();
+      return aTime - bTime;
+    });
+    const oldestPool = sorted[0];
+    if (
+      oldestPool &&
+      oldestPool.attributes &&
+      oldestPool.attributes.pool_created_at
+    ) {
+      return new Date(oldestPool.attributes.pool_created_at).getTime();
+    }
+    return undefined;
+  }
+
+  /**
+   * Validate the onchain response for a token
+   * @param data - The token data
+   * @param pools - The token pools
+   * @returns The token info
+   */
+  private parseOnchainResponse(response: TokenGetAddressResponse): TokenInfo {
+    const {
+      success,
+      error,
+      data: parsedData,
+    } = OnchainResponseSchema.safeParse(response);
+    if (!success) {
+      throw new Error(
+        `Invalid CoinGecko response for ${response.data?.attributes?.address || "token"}: ${error}`,
+      );
+    }
+    const { data, included: pools } = parsedData;
+    const { symbol, price_usd, volume_usd, total_reserve_in_usd, fdv_usd } =
+      data.attributes;
+    return {
+      price: parseFloat(price_usd),
+      symbol: symbol.toUpperCase(),
+      pairCreatedAt: this.getCreatedAtTimestampFromPools(pools),
+      volume: { h24: parseFloat(volume_usd.h24) },
+      liquidity: { usd: parseFloat(total_reserve_in_usd) },
+      fdv: parseFloat(fdv_usd),
+    };
+  }
+
+  /**
+   * Fetch price data from CoinGecko onchain API with retry logic
+   * @param tokenAddress - The token contract address (expected to be normalized as canonical base58 or checksummed)
+   * @param network - The CoinGecko network identifier (e.g., "eth", "solana")
+   * @returns The token info
+   */
+  private async fetchPrice(
+    tokenAddress: string,
+    network: CoinGeckoNetwork,
+  ): Promise<TokenInfo | null> {
+    try {
+      // Note: the CoinGecko SDK natively handles retries upon internal network failures. However,
+      // in case of 429 errors, `withRetry` will catch and retry manually with exponential backoff.
+      const getTokenInfo = () =>
+        this.client.onchain.networks.tokens.getAddress(tokenAddress, {
+          network,
+          include_composition: true,
+          include: "top_pools",
+        });
+      const response = await withRetry(getTokenInfo, {
```

Comment:
here's that `withRetry` helper and the _somewhat_ double retry logic...only for 429s, though, since the CG SDK should first do retries on 500s etc

(i'm debating if we should only use our helper, or if this dual-setup is better)